runtime/darwinia/src/messages/crab_messages.rs

```diff
@@ -0,0 +1,276 @@
+pub type ToCrabMessageVerifier<R> = DarwiniaFromThisChainMessageVerifier<WithCrabMessageBridge, R>;
+/// Message payload for Crab -> Darwinia messages.
+pub type FromCrabMessagePayload = FromBridgedChainMessagePayload<WithCrabMessageBridge>;
+/// Encoded Darwinia Call as it comes from Crab.
+pub type FromCrabEncodedCall = FromBridgedChainEncodedMessageCall<Call>;
+/// Messages proof for Crab -> Darwinia messages.
+type FromCrabMessagesProof = FromBridgedChainMessagesProof<Hash>;
+/// Messages delivery proof for Darwinia -> Crab messages.
+type ToCrabMessagesDeliveryProof = FromBridgedChainMessagesDeliveryProof<Hash>;
+/// Call-dispatch based message dispatch for Crab -> Darwinia messages.
+pub type FromCrabMessageDispatch =
+	FromBridgedChainMessageDispatch<WithCrabMessageBridge, Runtime, Ring, S2sBridgeDispatch>;
+
+/// Initial value of `CrabToDarwiniaConversionRate` parameter.
+pub const INITIAL_CRAB_TO_DARWINIA_CONVERSION_RATE: FixedU128 =
+	FixedU128::from_inner(FixedU128::DIV);
+
+frame_support::parameter_types! {
+	/// Crab to Darwinia conversion rate. Initially we treat both tokens as equal.
+	pub storage CrabToDarwiniaConversionRate: FixedU128 = INITIAL_CRAB_TO_DARWINIA_CONVERSION_RATE;
+}
+
+/// Darwinia -> Crab message lane pallet parameters.
+#[derive(Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug)]
+pub enum DarwiniaToCrabMessagesParameter {
+	/// The conversion formula we use is: `DarwiniaTokens = CrabTokens * conversion_rate`.
+	CrabToDarwiniaConversionRate(FixedU128),
+}
+impl MessagesParameter for DarwiniaToCrabMessagesParameter {
+	fn save(&self) {
+		match *self {
+			DarwiniaToCrabMessagesParameter::CrabToDarwiniaConversionRate(ref conversion_rate) => {
+				CrabToDarwiniaConversionRate::set(conversion_rate)
+			}
+		}
+	}
+}
+
+/// Darwinia <-> Crab message bridge.
+#[derive(Clone, Copy, RuntimeDebug)]
+pub struct WithCrabMessageBridge;
+impl MessageBridge for WithCrabMessageBridge {
+	const RELAYER_FEE_PERCENT: u32 = 10;
+	const THIS_CHAIN_ID: ChainId = DARWINIA_CHAIN_ID;
+	const BRIDGED_CHAIN_ID: ChainId = CRAB_CHAIN_ID;
+	const BRIDGED_MESSAGES_PALLET_NAME: &'static str = WITH_DARWINIA_MESSAGES_PALLET_NAME;
+
+	type ThisChain = Darwinia;
+	type BridgedChain = Crab;
+
+	fn bridged_balance_to_this_balance(bridged_balance: Balance) -> Balance {
+		Balance::try_from(CrabToDarwiniaConversionRate::get().saturating_mul_int(bridged_balance))
+			.unwrap_or(Balance::MAX)
+	}
+}
+
+/// Darwinia chain from message lane point of view.
+#[derive(Clone, Copy, RuntimeDebug)]
+pub struct Darwinia;
+impl messages::ChainWithMessages for Darwinia {
+	type Hash = Hash;
+	type AccountId = AccountId;
+	type Signer = AccountPublic;
+	type Signature = Signature;
+	type Weight = Weight;
+	type Balance = Balance;
+}
+impl messages::ThisChainWithMessages for Darwinia {
+	type Call = Call;
+
+	fn is_outbound_lane_enabled(lane: &LaneId) -> bool {
+		*lane == [0, 0, 0, 0] || *lane == [0, 0, 0, 1] || *lane == DARWINIA_CRAB_LANE
```

Comment:
```suggestion
		*lane == DARWINIA_CRAB_LANE
```

DARWINIA_CRAB_LANE is `[0, 0, 0, 0]` in current runtime.